src/Npgsql/ConnectorPool.cs

```diff
@@ -360,7 +360,8 @@ internal void Return(NpgsqlConnector connector)
             // If Clear/ClearAll has been been called since this connector was first opened,
             // throw it away. The same if it's broken (in which case CloseConnector is only
             // used to update state/perf counter).
-            if (connector.ClearCounter < _clearCounter || connector.IsBroken)
+            
+            if (connector.ClearCounter < _clearCounter || connector.IsBroken || connector.IsAppropriateFor(connector.Settings.TargetServerType) == false)
```

Comment:
Ah yes, the server can change type at any point during the connection lifetime; so just because you initially connected to a standby, doesn't mean it will remain one over time so without verifying this when being handed back to the pool you end up with a 'poisoned' connection pool for a given connectionstring

Reply:
It definitely can change, but wouldn't `KeepAlive` in this case just break the connection?

Reply:
Yes it should. I have no idea why this is here. I'll take it out and retest my scenario's in-case there's something critical